source/core_sntp_serializer.c

```diff
@@ -191,6 +203,87 @@ static uint32_t readWordFromNetworkByteOrderMemory( const uint32_t * ptr )
+ * between server and client timestamps represents that the server and client
+ * are within 34 years of each other to be able to calculate the clock-offset
+ * value according to the NTPv4 specification's on-wire protocol.
+ *
+ * @param[in] firstOrderDiff A first-order difference value between the client
+ * and server.
+ *
+ * @note As the SNTP timestamp value wraps around after ~136 years (exactly at
+ * 7 Feb 2036 6h 28m 16s), the utility logic checks the first order difference
+ * in both polarities (i.e. as (Server - Client) and (Client - Server) time values )
+ * to support the edge case when the two timestamps are in different SNTP eras (for
+ * example, server time is in 2037 and client time is in 2035 ).
+ */
+static bool isEligibleForClockOffsetCalculation( uint32_t firstOrderDiff )
+{
+    /* Note: The (UINT32_MAX  - firstOrderDiff + 1U) expression represents
+     * 2's complement or negation of value.
+     * This is done to be compliant with both CBMC and MISRA Rule 10.1.
+     * CBMC flags overflow for (unsigned int = 0U - positive value) whereas
+     * MISRA rule forbids use of unary minus operator on unsigned integers. */
+    return( ( ( firstOrderDiff & CLOCK_OFFSET_FIRST_ORDER_DIFF_OVERFLOW_BITS_MASK ) == 0U ) ||
+            ( ( ( UINT32_MAX - firstOrderDiff + 1U ) & CLOCK_OFFSET_FIRST_ORDER_DIFF_OVERFLOW_BITS_MASK ) == 0U ) );
+}
+
+/**
+ * @brief Utility to perform a safe subtraction operation of unsigned integers and
+ * return the value as a signed integer. This function returns the effective subtraction
+ * value as ( @p minuend - @p subtrahend ).
+ *
+ * @note This utility provides safe subtraction result that involve the following 2 operations::
+ *  * Safe subtraction between unsigned integers
+ *                    AND
+ *  * Safe conversion of unsigned integer to signed integer
+ *
+ * @param[in] minuend The value to subtract from.
+ * @param[in] subtrahend The amount of value to subtract from @p minuend.
+ *
+ * @return The calculated signed subtraction value between the unsigned integers.
+ */
+static int32_t safeSubtraction( uint32_t minuend,
+                                uint32_t subtrahend )
+{
+    int32_t calculatedValue = 0;
+
+    /* The correct polarity of subtraction is "minuend - subtrahend",
+    * but to avoid overflow in subtraction of unsigned integers, we perform
+    * subtraction in the polarity that generates a positive value. */
+    bool polarity = ( minuend > subtrahend ) ? true : false;
+    uint32_t positiveDiff = ( polarity == true ) ? minuend - subtrahend :
+                            subtrahend - minuend;
+
+    /* Check whether the difference value has the most significant bit set.
+     * An unsigned integer with the most significant bit set cannot be safely assigned
+     * to a signed integer.*/
+    if( ( positiveDiff & UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK ) == UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK )
```

Comment:
[Nitpick]

Would it be cleaner to check `positiveDiff > INT32_MAX` instead? Basically, we would be explicit about the overflow; we can mention that in the comment too -- that the reason we have this `if` branch is to handle overflows correctly.

If we choose to make this change, we can also remove the definition for `UINT32_MOST_SIGNIFICANT_BIT_BIT_MASK`.

Reply:
Yes, done that!